src/bin/merchant/pay.rs

```diff
@@ -46,37 +71,41 @@ impl Pay {
             }
         };
 
-        proceed!(in chan);
+    proceed!(in chan);
 
-        // Run the zkAbacus.Pay protocol
-        let pay_result = zkabacus_pay(rng, database, session_key, chan, payment_amount)
-            .await
-            .context("Payment failed");
-
-        match pay_result {
-            Ok(chan) => {
-                // Send the response note (i.e. the fulfillment of the service) and close the
-                // connection to the customer
-                let response_note = approve::payment_success(client, response_url).await;
-                let (note, result) = match response_note {
-                    Err(err) => (None, Err(err)),
-                    Ok(o) => (o, Ok(())),
-                };
-                chan.send(note)
-                    .await
-                    .context("Failed to send response note")?
-                    .close();
-                result
-            }
-            Err(err) => {
-                approve::failure(client, response_url).await;
-                Err(err)
-            }
+    Ok((response_url, chan))
+}
+
+/// Inform the approver service whether the payment succeeded and pass the resulting fulfillment
+/// to the customer.
+async fn provide_service(
+    response_url: Option<Url>,
+    maybe_chan: Result<Chan<pay::MerchantProvideService>, anyhow::Error>,
+    client: &reqwest::Client,
+) -> Result<(), anyhow::Error> {
+    match maybe_chan {
+        Ok(chan) => {
+            // Send the response note (i.e. the fulfillment of the service) and close the
+            // connection to the customer
+            let response_note = approve::payment_success(client, response_url).await;
+            let (note, result) = match response_note {
+                Err(err) => (None, Err(err)),
+                Ok(o) => (o, Ok(())),
+            };
+            chan.send(note)
+                .await
+                .context("Failed to send response note")?
+                .close();
+            result
+        }
+        Err(err) => {
+            approve::failure(client, response_url).await;
+            Err(err)
         }
```

Comment:
Why extract this into its own function if it's only going to be called once?

Reply:
Same as above, I prefer to have the ugly details grouped into a single high-level, easily documented function. This one is a bit less necessary since it's only statement, but that's why.